### DISCUSSION TITLE
Install local homebrew formulae with custom prefix

### DISCUSSION THREAD

**Discussion by ozerianskyycoro** [Homebrew/discussions#4702] [Casks] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.1.4-20-gdc42544
ORIGIN: https://github.com/Homebrew/brew
HEAD: dc42544c59a26805825a56b323caeb9bbf696ff3
Last commit: 19 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 908f5efde15cb263ace7fc388ec9fff4f6e5c6c7
Core tap last commit: 3 hours ago
Core tap branch: master
Core tap JSON: 09 Aug 09:07 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 16
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 16-core 64-bit kabylake
Clang: 14.0.0 build 1400
Git: 2.37.1 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 13.5-x86_64
CLT: 14.3.0.0.1.1679647830
Xcode: 14.1
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some taps are not on the default git origin branch and may not receive
updates. If this is a surprise to you, check out the default branch with:
  git -C $(brew --repo ozerianskyycoro/tap) checkout 

Warning: Your Xcode (14.1) is outdated.
Please update to Xcode 14.3 (or delete it).
Xcode can be updated from the App Store.


Warning: Broken symlinks were found. Remove them with `brew cleanup`:
  /usr/local/lib/ruby/gems/3.2.0/bin/bundle
  /usr/local/lib/ruby/gems/3.2.0/bin/bundler
  /usr/local/opt/python@3
```

### Description of issue

**Description of Issue:**
Hello Homebrew community,

I'm currently facing an issue while trying to install a local formula named `openvpn.rb`. I want to install it with custom prefix, so openvpn.dylib will look for dependencies inside this prefix and not in `usr/local`. Despite my attempts of running edited formulae with `brew install ./openvpn.rb` it's still install open vpn and all dependencies in `usr/local`

**Steps I've Taken:**
- Cloned the repository and navigated to the formula's directory.
- Edited the formula file to ensure it follows the correct syntax and structure.
- Tried to install the formula using `brew install ./openvpn.rb`, but encountered the error.
- Uninstalled the existing OpenVPN installation to start fresh.

**Formula Contents (openvpn.rb):**
```
class Openvpn < Formula
  desc "SSL/TLS VPN implementing OSI layer 2 or 3 secure network extension"
  homepage "https://openvpn.net/community/"
  url "https://swupdate.openvpn.org/community/releases/openvpn-2.6.5.tar.gz"
  mirror "https://build.openvpn.net/downloads/releases/openvpn-2.6.5.tar.gz"
  sha256 "e34efdb9a3789a760cfc91d57349dfb1e31da169c98c06cb490c6a8a015638e2"
  license "GPL-2.0-only" => { with: "openvpn-openssl-exception" }

  livecheck do
    url "https://openvpn.net/community-downloads/"
    regex(/href=.*?openvpn[._-]v?(\d+(?:\.\d+)+)\.t/i)
  end

  bottle do
    sha256 arm64_ventura:  "8bbc0fc0eac54e799776419cdf9958f669c46fb550a9b8d865a3efda4afcd962"
    sha256 arm64_monterey: "2324f818dc4c1d6c803a84423bc4a03bb82f17bd2323c4847753450d4a5c4858"
    sha256 arm64_big_sur:  "c3979e83364ca831f333397bed143974c70e4a8f544eafa4075feec14c6d414f"
    sha256 ventura:        "5688988f1132f2b0505626c9a921fe72275234cd3a79a88bf0869a0c46b53aab"
    sha256 monterey:       "8a66be9610feeffdf01a5b576ddd89a1e5fe2996f2202d2c2a5fb897a6e76c6d"
    sha256 big_sur:        "71978a23eec4f0ed94ff147eaadcf48f8cd6085737a9a1205d30f97d1a8f2474"
    sha256 x86_64_linux:   "0e1ee5e55b05a4aba7f452e2da9dfc41a27203e75c9941614eac5eb9e6a546e8"
  end

  depends_on "pkg-config" => :build
  depends_on "lz4"
  depends_on "lzo"
  depends_on "openssl@3"
  depends_on "pkcs11-helper"

  on_linux do
    depends_on "libcap-ng"
    depends_on "libnl"
    depends_on "linux-pam"
    depends_on "net-tools"
  end

  def install
    system "./configure", "--disable-debug",
                          "--disable-dependency-tracking",
                          "--disable-silent-rules",
                          "--with-crypto-library=openssl",
                          "--enable-pkcs11",
                          "--prefix=/My/Path"
    inreplace "sample/sample-plugins/Makefile" do |s|
      if OS.mac?
        s.gsub! Superenv.shims_path/"pkg-config", Formula["pkg-config"].opt_bin/"pkg-config"
      else
        s.gsub! Superenv.shims_path/"ld", "ld"
      end
    end
    system "make", "install"

    inreplace "sample/sample-config-files/openvpn-startup.sh",
              "/etc/openvpn", etc/"openvpn"

    (doc/"samples").install Dir["sample/sample-*"]
    (etc/"openvpn").install doc/"samples/sample-config-files/client.conf"
    (etc/"openvpn").install doc/"samples/sample-config-files/server.conf"

    # We don't use mbedtls, so this file is unnecessary & somewhat confusing.
    rm doc/"README.mbedtls"
  end

  def post_install
    (var/"run/openvpn").mkpath
  end

  service do
    run [opt_sbin/"openvpn", "--config", etc/"openvpn/openvpn.conf"]
    keep_alive true
    require_root true
    working_dir etc/"openvpn"
  end

  test do
    system sbin/"openvpn", "--show-ciphers"
  end
end

```

### Relevant casks

openvpn

**Comment by SMillerDev** [comment, marked as answer]
Homebrew doesn't install things outside of it's own prefix so this does not seem possible.

**Comment by osalbahr** [comment]
> I'm currently facing an issue while trying to install a local formula named openvpn.rb. I want to install it with custom prefix, so openvpn.dylib will look for dependencies inside this prefix and not in usr/local. Despite my attempts of running edited formulae with brew install ./openvpn.rb it's still install open vpn and all dependencies in usr/local

You can technically try to install in the custom prefix via the [unsupported method](https://docs.brew.sh/Installation#untar-anywhere-unsupported) but you’d have to manually put the custom dependencies in Homebrew’s directories which can get messy. 

You’d probably want to create a [Tap](https://docs.brew.sh/Taps), which is supported, instead of trying to force Homebrew’s `openvpn` to use different dependencies.

**Comment by nanoscopic** [comment]
I've created two projects that are capable of installing brew bottles wherever you want and making them work together.

If you either have or can get the bottles for the packages you want, you can use what I've made to install brew packages anywhere on macos intel/arm.

See https://github.com/dryark/minibrew_deploy as an example of deploying a collection of bottles.
This repo can be cloned and 'deploy.pl' run to extract the bottles detailed in dlinfo.json and set them up to work together.

dlinfo.json and dlsize.json were created using https://github.com/dryark/minibrew ; still refining the process of how all this works, but it is meant to let you avoid the complexity of brew for packages that have bottles available and postinstall steps can be either ignored or done manually.
